### PR TITLE
Refactor packaging configuration, setup, and testing

### PR DESCRIPTION
**Pull request by tepickering**
This PR does a major refactoring of the packaging setup, configuration, installation, and testing. It leverages the capabilities that `setuptools`, `pip` , and `tox` now provide to enable dependency handling at install-time, integrate package versioning with git tagging, and enable testing within isolated virtual environments. The most major changes include:

- Move package configuration information from `setup.py` to `setup.cfg`. This makes `setup.py` a fairly simple piece of boilerplate while `setup.cfg` provides a general configuration file for both package information as well as tools used by the package (e.g. `pytest` and `coverage`). Moving the package configuration into `setup.cfg` makes it a lot more straightforward to define groups of optional dependencies. Obvious use cases are to define such groups for doing testing and building the documentation. Other groups are added here to manage installation of `pyqt5`, `pyside2`, and `shapely`. 
- Many of the scripts in `PypeIt/bin` were simple wrappers to scripts in `PypeIt/pypeit/scripts`. Those scripts have been deprecated and replaced by `entry_points` in `setup.cfg` that point directly to the files in `PypeIt/pypeit/scripts`. For each of those cases I needed to add an `entry_point()` method to make this possible because entry points can't take arguments. The `entry_point()` methods basically do exactly what the old wrapper scripts did. The way I did this violates the principle of DRY, but is at least minimally invasive. I have ideas for more clever ways to do this, but that is more appropriate for a future PR. The scripts in `PypeIt/bin` that aren't simple wrappers have been left as-is and defined as `scripts` as before.
- Implement `setuptools_scm` to automate versioning based on git tags. It uses the latest tag in the history to generate a version number and automatically increments and adds the `dev` part if there are commits after the latest tag. This is written to `pypeit/version.py` and the `bumpversion` config was followed to make the appropriate code changes to utilize the generated version. The notable places are `pypeit/__init__.py`, `doc/conf.py`, and `doc/index.rst`. As discussed on slack, we will want to clean up the existing tags before this is merged so that the versions generated by `setuptools_scm` will match what we want.  
- Implement `tox` to provide the ability to run tests within isolated environments. This allows testing of the code along with its dependencies and control what versions of each dependency, including python version, are installed for each test. It is still possible to run tests manually by simply doing `pytest`. However, the move to `setup.cfg` got rid of `python setup.py test`. That now prints a usage message. Since all that did before was run `pytest`, I figure that's not much of a loss. By default `tox` uses `pip` and `virtualenv` to handle dependencies and environments. There is a `tox-conda` package that allows the use of `conda` instead. A test environment is defined to use this along with our `environment.yml` file to run tests under `conda`.
- The CI has been refactored to use the `tox` tests so that testing locally and testing via CI should more exactly match. The matrix of `tox` environments tested includes the `conda` environment. A second matrix is defined to run basic tests on `macos` and `windows`. The `windows` test is currently failing due to issues relating to `memmap` and FITS files. 
-  Cron schedules have been added to the CI so that tests on `master` and `develop` are run on a weekly basis. The schedule for `master` is added to the existing CI workflow. Running a scheduled workflow on a non-default branch required adding a new workflow that specifically checks that branch out. This was done to do scheduled tests on `develop`. Those tests are currently set up to only run on `ubuntu-latest`, but the other OS's can easily be added. The motivation is mostly to catch issues from upstream development of various dependencies. 

Other smaller changes/fixes are included as well. Notably, the scheme for getting the username to check against a list of developers was breaking under windows and caught by the new CI test. I patched that and it's happy now. Lots of documentation updates got added when i ran `update_docs` to reflect things that got deprecated and to test whether my changes to version handling flowed through properly. 

I also made significant changes to `doc/installing.rst`  and `doc/dev/development.rst` to reflect the new functionality and capabilities.

### CODE REVIEW

**Review by profxj — APPROVED** [review]
Wow, super helpful!  I must admit I didn't give it a very careful review, but the proof will be in the pudding.  Am approving and will let someone else test the pudding. :)

**Comment by tepickering** [comment]
btw, an easy way to give this a spin is to directly install from this PR via pip:
```
pip install "git+https://github.com/pypeit/PypeIt.git@refs/pull/1130/head#egg=pypeit[test,docs,pyqt5]"
```
i did need to make sure to `apt install llvm` was done in my linux install so that numba would work. i guess the github actions runner must already have it because there have been no CI complaints.

File: MANIFEST.in
Comment (kbwestfall) [inline MANIFEST.in:14]: We should also delete/deprecate the requirements_doc.txt file at the top-level.

File: setup.cfg
Comment (kbwestfall) [inline setup.cfg:52]: I see you just moved this file, but can we rename it?  Something like `pypeit_cp_spec1d`, including the removal of the `.py` extension.  I prefer executable scripts to not have the extension.

File: doc/api/pypeit.rst
Comment (kbwestfall) [inline doc/api/pypeit.rst:30]: I'm not sure why this is in the file twice.  This file should be automatically generated by the `update_docs` script.  If you didn't edit this by hand, I can check why this is occurring twice.

File: doc/api/pypeit.rst
Comment (kbwestfall) [inline doc/api/pypeit.rst:52]: This probably should be removed.  I think you can have sphinx ignore the file by editing line 59 of `doc/Makefile` to:

```
	SPHINX_APIDOC_OPTIONS=$(SPHINXAPIOPT) $(SPHINXAPI) --separate -o ./api ../pypeit ../pypeit/tests/* ../pypeit/scripts/* ../pypeit/deprecated/* ../pypeit/version.py
```

File: doc/conf.py
Comment (kbwestfall) [inline doc/conf.py:24]: Nice.

File: doc/help/pypeit_chk_alignments.rst
Comment (kbwestfall) [inline doc/help/pypeit_chk_alignments.rst:9]: It's annoying that these files get changed based on the width of the terminal at the instance when the docs are updated.  I'll look into fixing the width to 80 characters for these output files.

File: doc/index.rst
Comment (kbwestfall) [inline doc/index.rst:41]: Sorry if I missed it, but where is this defined?

File: doc/installing.rst
Comment (kbwestfall) [inline doc/installing.rst:73]: Thanks for pointing this out.  This actually encouraged me to switch from `pyenv` to `virtualenv`.  Here, we might also suggest using `virtualenvwrapper`: https://virtualenvwrapper.readthedocs.io/en/latest/ .

File: doc/installing.rst
Comment (kbwestfall) [inline doc/installing.rst:154]: Same questions as in the development doc: Do we need both `pyqt5` and `pyside2` in this call, and can we consolidate all of these (as well as keeping them separate) into a `dev` option?

File: doc/installing.rst
Comment (kbwestfall) [inline doc/installing.rst:174]: Hopefully we can remove this after a @jhennawi PR that removes the numba dependence.

File: doc/installing.rst
Comment (kbwestfall) [inline doc/installing.rst:194]: Does `pip install -e .` not compile the C code?

File: pypeit/pypmsgs.py
Comment (kbwestfall) [inline pypeit/pypmsgs.py:64]: Not really relevant to this PR, but my opinion is that we should deprecate this logging functionality.

**Review by kbwestfall — CHANGES_REQUESTED** [review]
Awesome, @tepickering !!  I checked out your fork and gave things a test run.  I had a few comments that are mostly minor.  My only requested change is that you have a look at the `tox` failure that I ran into.  Thanks!

File: doc/dev/development.rst
Comment (tepickering) [reply doc/dev/development.rst:35]: i dug around to see if there's a clever way to do this, but there isn't. at least not yet; it's a requested feature. so i brute forced it and made a `dev` section that's copy-pastaed from the other sections.

File: doc/dev/development.rst
Comment (tepickering) [reply doc/dev/development.rst:498]: that's because you're pulling off my fork which doesn't have the same tags as the main repo. i noticed this same thing. it's correct given the latest tag in my fork and i played around with tags in various checkouts to make sure `setuptools_scm` is doing the right thing and it is. so once this is merged to `master` and a tag of `1.3.1` set there, it should all be gtg.

File: .github/workflows/ci_cron.yml
Comment (tepickering) [reply .github/workflows/ci_cron.yml:34]: yes, they'll show up under the actions tab once they're run. and emails will go out if they fail. however, iirc they won't start triggering until they're merged.

File: doc/dev/development.rst
Comment (tepickering) [reply doc/dev/development.rst:498]: the back story is that my fork is quite old. i can try to update the tags there, if you want. probably want to do the 1.3 -> 1.3.0 fix first, though.

File: MANIFEST.in
Comment (tepickering) [reply MANIFEST.in:14]: yes, good point. i need to refactor readthedocs.yml as well to match that. was on my list and missed it. it's mostly boilerplate so it's pretty easy.

File: doc/dev/development.rst
Comment (tepickering) [reply doc/dev/development.rst:340]: i take it you have `PYPEIT_DEV` set, correct? because i think if you don't, that test gets skipped. so try doing `tox -e test-pyside2-shapely` so that you have the same env as the one `pytest` is using.

File: setup.cfg
Comment (tepickering) [reply setup.cfg:52]: agreed. renamed the script and updated `setup.cfg`.

File: doc/api/pypeit.rst
Comment (tepickering) [reply doc/api/pypeit.rst:30]: i did not edit any of the api docs by hand.

File: doc/api/pypeit.rst
Comment (tepickering) [reply doc/api/pypeit.rst:52]: good catch. made this change.

File: doc/help/pypeit_chk_alignments.rst
Comment (tepickering) [reply doc/help/pypeit_chk_alignments.rst:9]: if you can find the trick, i can try folding it in here and re-generating the docs so there aren't so many changes.

File: doc/index.rst
Comment (tepickering) [reply doc/index.rst:41]: that gets defined in `conf.py` and `sphinx` gets it from there. i just followed what the comment in `conf.py` said and it worked.

File: doc/installing.rst
Comment (tepickering) [reply doc/installing.rst:73]: i use `conda` personally, but i'll check that out and update the docs.

File: doc/installing.rst
Comment (tepickering) [reply doc/installing.rst:154]: now that i have `dev` set up, i'll update this to use that.

File: doc/installing.rst
Comment (tepickering) [reply doc/installing.rst:194]: it should. if not, i would expect the tests to catch that. not sure why this got changed back to the old wording...

File: pypeit/pypmsgs.py
Comment (tepickering) [reply pypeit/pypmsgs.py:64]: yes, it seemed a bit janky to me, but out-of-scope for this PR.

File: MANIFEST.in
Comment (tepickering) [reply MANIFEST.in:14]: requirements_doc.txt is now deprecated and readthedocs.yml updated to install via pip and include the `docs` extra requirements.

File: doc/dev/development.rst
Comment (kbwestfall) [reply doc/dev/development.rst:35]: Nice.  Thanks for that.

File: doc/help/pypeit_chk_alignments.rst
Comment (kbwestfall) [reply doc/help/pypeit_chk_alignments.rst:9]: I found the trick, but it's a bit annoying.  For each script, you have to change the formatter to something like the following:
```
    width = 78 if return_parser else None
    parser = argparse.ArgumentParser(description=run_pypeit_usage(),
                                     formatter_class=lambda prog:
                                         argparse.RawDescriptionHelpFormatter(prog, width=width))
```
That's for `run_pypeit`, but the details of the formatter class sometimes changes.  I suggest we leave this fix to a latter PR.

File: doc/index.rst
Comment (kbwestfall) [reply doc/index.rst:41]: Nice.

File: doc/dev/development.rst
Comment (kbwestfall) [reply doc/dev/development.rst:340]: Nice.  That fixed it.  Thanks!

File: doc/dev/development.rst
Comment (kbwestfall) [reply doc/dev/development.rst:498]: I vote that we just merge this PR now and then hopefully you can help me fix the version numbering when it makes it into develop.

File: doc/api/pypeit.rst
Comment (kbwestfall) [reply doc/api/pypeit.rst:30]: Weird.  I just re-ran `update_docs`, and this line wasn't repeated.  I think it's fine, and we can fix it in `develop` if need be.

**Review by kbwestfall — APPROVED** [review]
Thanks for all the fixes.  I think this is ready to be merged!  I tried to re-run the CI tests, and they show that they finished find (again).  But they're still showing up as not finished, which will block the merge.  We may have to have @profxj force it...

File: doc/help/pypeit_chk_alignments.rst
Comment (tepickering) [reply doc/help/pypeit_chk_alignments.rst:9]: ah, ok. thanks for looking into it.

**Comment by tepickering** [comment]
looks like the tests i added are working fine and pass. the ones that are hanging are the tests from the workflow that is being outdated by this PR. i will look into working around this without having to force it.

**Comment by tepickering** [comment]
this looks like the recipe to removing the stale checks: https://stackoverflow.com/questions/64210553/deleting-stale-checks-on-github

File: doc/installing.rst
Comment (jhennawi) [inline doc/installing.rst:178]: I think this should be (or PySide2) not (in addition to PySide2) right?

Comment (tepickering) [reply doc/installing.rst]: yeah, i reworded this to generalize it to a dependency on QT bindings.

File: doc/installing.rst
Comment (tepickering) [reply doc/installing.rst:150]: good point. added that.

**Comment by tepickering** [comment]
thanks for the reviews! i think i incorporated all the requested changes. merging is still blocked by the stale CI checks due to the test names being changed.

**Comment by jhennawi** [comment]
Hi @tepickering, 

One comment. I successfully installed on centOS linux after removing the numba dependency from my branch, which is now the case. One minor point though, when I tried to test I got the following :

Screenshot 2021-01-31 at 17 51 30

This is also good, but you should update the "Test Your Installation" section of installing.rst to reflect the new method for testing since I was just following that, i.e. we have the deprecated instructions there not the new way of testing with tox, etc. 

Thanks!
Joe

**Comment by tepickering** [comment]
ah, thanks for pointing that out. i missed that and will update it accordingly. i'll include the nuance of when/why you should use `tox` vs `pytest`.

**Comment by jhennawi** [comment]
One other minor point @tepickering, the "For Developers" link installing.rst leads nowhere. You might want to also correct that.  Thanks. I suppose also this testing info regarding tox should be both in the installing doc and the dev doc. 

Screenshot 2021-02-01 at 15 44 50

**Pull request merged**
